python/lsst/daf/butler/remote_butler/server_models.py

```diff
@@ -141,3 +154,93 @@ class QueryCollectionsResponseModel(pydantic.BaseModel):
 
     collections: list[str]
     """Collection names that match the search."""
+
+
+class MaterializedQuery(pydantic.BaseModel):
+    """Captures the parameters from a call to ``QueryDriver.materialize``."""
+
+    type: Literal["materialized"] = "materialized"
+    key: UUID
+    tree: SerializedQueryTree
+    dimensions: SerializedDimensionGroup
+    datasets: list[str]
+
+
+class DataCoordinateUpload(pydantic.BaseModel):
+    """Captures the parameters from a call to
+    ``QueryDriver.upload_data_coordinates``.
+    """
+
+    type: Literal["upload"] = "upload"
+    key: UUID
+    dimensions: SerializedDimensionGroup
+    rows: list[list[DataIdValue]]
+
+
+AdditionalQueryInput: TypeAlias = Annotated[
+    MaterializedQuery | DataCoordinateUpload, pydantic.Discriminator("type")
+]
+"""Information about additional data tables that may be used by a query."""
+
+
+class QueryInputs(pydantic.BaseModel):
+    """Serialized Butler query with additional context needed to execute it."""
+
+    tree: SerializedQueryTree
+    default_data_id: SerializedDataCoordinate
```

Comment:
Really the only rough edge in the separation of concerns is that the default data ID had to passed to the server here so the DirectQueryDriver can use it.

Not really a big deal, and many of the other Butler methods have a similar issue. When I implement default data ID for RemoteButler basically every method that uses `findDataset` internally is going to need the default data ID sent like this.

Reply:
> But we'll eventually have per-user query state that will have to be persistent server-side (materializations, data ID uploads, and result pages)

I must have missed something because I thought we were wanting the server to be mostly stateless. Query workers would be running and have an ID that would be used by the client to find what the worker state was but the server wouldn't be keeping track of it itself internally -- there would not be a worker queue per server instance. I thought that temp tables in specific database servers known to specific butler servers was difficult unless you use a websocket or something to keep the connection from the client pinned to a specific server?

Reply:
I think this is in the space between "mostly" and "completely" stateless.

I had imagined temp tables being emulated via a reverse of the parquet-pagination approach, where we'd dump them to a file on the server and re-upload to the database as needed, but I'm not sure that actually makes the most sense - or even any sense at all, if it's a bigger hopper from server to DB than it is from client to server.  Regular database tables that we clean up manually may be another option, and `raise NotImplementedError` might be, too.

We've also talked about spawning a new process for each query, and while that wouldn't solve this problem directly, I could imagine us spawning a new process at the start of each user query context instead if we find the overheads to be a problem, and regardless it's another kind of user-specific resource the server has to manage the lifetime of.

All of these things are difficult, and I'm all for putting them off until we have basic functionality working.

Reply:
The easiest way they would work is basically as Jim described -- they are stored as files in S3, there is an entry in redis so we can expire them later, and that redis entry is tied to the UUID returned to the client for use in its QueryTree.   

It's easiest if the query worker pool re-uploads the data to the DB for each query execution.  In theory we might have the query workers maintain an LRU cache of already-uploaded tables and preferentially route queries to a  database instance that has the needed tables already.

Reply:
i.e. each individual server process is stateless but there is global state for the pool stored in Redis.